lib/gc/Transforms/MergeAlloc.cpp

```diff
@@ -0,0 +1,352 @@
+                       [](const TraceWithTick &a, const TraceWithTick &b) {
+                         return a.tick < b.tick;
+                       });
+      auto &retTrace = ret[scope];
+      retTrace.reserve(trace.size());
+      for (auto &tr : trace) {
+        retTrace.emplace_back(tr.trace);
+      }
+    }
+    return ret;
+  }
+};
+
+} // namespace
+
+FailureOr<mlir::gc::memoryplan::ScopeTraceData>
+collectMemoryTrace(Operation *root,
+                   const mlir::BufferViewFlowAnalysis &aliasAnaly,
+                   bool markOnly) {
+  TickCollecter collecter{aliasAnaly};
+  root->walk<WalkOrder::PreOrder>([&](Operation *op) {
+    collecter.popScopeIfNecessary(op);
+    collecter.forwardTick();
+    if (auto viewop = dyn_cast<ViewLikeOpInterface>(op)) {
+      collecter.onMemrefViews(viewop);
+    } else if (op->hasTrait<OpTrait::ReturnLike>()) {
+      collecter.onReturnOp(op);
+    } else if (!isMemoryEffectFree(op)) {
+      // if the op has no memory effects, it don't contribute to liveness
+      collecter.onGeneralOp(op);
+    }
+    // finally, if op is complex scope, push one ComplexScope
+    if (isa<RegionBranchOpInterface>(op) || isa<LoopLikeOpInterface>(op)) {
+      collecter.pushComplexScope(op);
+    }
+  });
+  collecter.popScopeIfNecessary(nullptr);
+  if (markOnly) {
+    for (auto &[alloc, tick] : collecter.allocTicks) {
+      auto allocscope = getAllocScope(alloc);
+      alloc->setAttr(
+          "__mergealloc_lifetime",
+          DenseI64ArrayAttr::get(root->getContext(),
+                                 {reinterpret_cast<int64_t>(allocscope),
+                                  tick.firstAccess, tick.lastAccess}));
+      allocscope->setAttr(
+          "__mergealloc_scope",
+          IntegerAttr::get(mlir::IntegerType::get(root->getContext(), 64),
+                           reinterpret_cast<int64_t>(allocscope)));
+    }
+    return mlir::gc::memoryplan::ScopeTraceData();
+  }
+  return collecter.getTrace();
+}
+
+} // namespace gc
+} // namespace mlir
+
+namespace {
+struct MergeAllocPass : mlir::gc::impl::MergeAllocBase<MergeAllocPass> {
+  using parent = mlir::gc::impl::MergeAllocBase<MergeAllocPass>;
+  void runOnOperation() override {
+    auto op = getOperation();
+    mlir::BufferViewFlowAnalysis aliasAnaly{op};
+    auto tracesOrFail =
+        mlir::gc::collectMemoryTrace(op, aliasAnaly, this->optionCheck);
+    if (failed(tracesOrFail)) {
+      signalPassFailure();
+      return;
+    }
+    if (this->optionCheck) {
+      return;
+    }
+    std::unordered_map<uintptr_t, std::vector<uintptr_t>> dummy;
+    for (auto &[scope, traces] : *tracesOrFail) {
+      std::unordered_map<uintptr_t, std::size_t> outSchedule;
+      if (traces.empty())
+        continue;
+      auto total = mlir::gc::memoryplan::schedule_memory_allocations(
+          traces, 64, !this->optionNoLocality,
+          gc::memoryplan::inplace_info_map(), outSchedule, dummy);
```

Comment:
Excuse me, may I ask - can “gc::memoryplan::inplace_info_map()” be automatically derived through static analysis?

Reply:
The memory allocation planner of current PR is extracted from our legacy code, which supports in-place operations. Currently, we don't support in-place allocation in the current design, but I keep the interface so that we may enable it in the future. By in-place operations, I mean operations like `Z = linalg.add X, Y`, where `X` can reuse buffer of `Z`. However, in the current implementation, `X, Y, Z` have overlapping lifetime.